File: C++/Trie.cpp
#include <string>
#include <iostream>
#define MAX 26
using namespace std;
struct node {
    struct node* child[MAX];
    int countLeaf;
};
struct node *newNode()
{
    struct node *Node = new struct node;
    Node->countLeaf = 0;
    for (int i = 0; i < MAX; i++)
    {
        Node->child[i] = NULL;
    }
    return Node;
}
void addWord(struct node *root, string s)
{
    int ch;
    struct node *temp = root;
    for (int i = 0; i < s.size(); i++)
    {
        ch = s[i] - 'a';
        if (temp->child[ch] == NULL)
        {
            temp->child[ch] = newNode();
        }
        temp = temp->child[ch];
    }
    temp->countLeaf++;
}
bool findWord(struct node *root, string s)
{
    int ch;
    struct node *temp = root;
    for (int i = 0; i < s.size(); i++)
    {
        ch = s[i] - 'a';
        if (temp->child[ch] == NULL)
        {
            return false;
        }
        temp = temp->child[ch];
    }
    return temp->countLeaf > 0;
}
bool isWord(struct node *pNode)
{
    return (pNode->countLeaf != 0);
}
bool isEmpty(struct node *pNode)
{
    for (int i = 0; i < MAX; i++)
    {
        if (pNode->child[i] != NULL)
        {
            return false;
        }
    }
    return true;
}
bool removeWord(struct node *root, string s, int level, int len)
{
    if (!root)
        return false;
    int ch = s[level] - 'a';
    if (level == len)
    {
        if (root->countLeaf > 0)
        {
            root->countLeaf--;
            return true;
        }
        return false;
    }
    int flag = removeWord(root->child[ch], s, level + 1, len);
    if (flag && !isWord(root->child[ch]) && isEmpty(root->child[ch]))
    {
        delete root->child[ch];
        root->child[ch] = NULL;
    }
    return flag;
}

File: Python3/Trie.py
class Node:
    def __init__(self):
        self.countLeaf = 0
        self.child = dict()
def addWord(root, s):
    tmp = root
    for ch in s:
        if ch in tmp.child:
            tmp = tmp.child[ch]
        else:
            tmp.child[ch] = Node()
            tmp = tmp.child[ch]
    tmp.countLeaf += 1
def findWord(root, s):
    tmp = root
    for ch in s:
        if ch not in tmp.child:
            return False
        tmp = tmp.child[ch]
    return tmp.countLeaf > 0

def isWord(node):
    return node.countLeaf != 0

def isEmpty(node):
    return len(node.child) == 0

def removeWord(root, s, level, len):
    if root == None:
        return False
    if level == len:
        if root.countLeaf > 0:
            root.countLeaf -= 1
            return True
        return False
    ch = s[level]
    flag = removeWord(root.child[ch], s, level + 1, len)
    if flag == True and isWord(root.child[ch]) == False and \
        isEmpty(root.child[ch]) == True:
            tmp = root.child[ch]
            del tmp
            del root.child[ch]
    return flag
